core/src/main/java/kafka/server/ClientMetricsManager.java

```diff
@@ -30,17 +68,366 @@ public class ClientMetricsManager implements Closeable {
+            if (instanceMetadata.isMatch(info.matchPattern())) {
+                allMetricsSubscribed = allMetricsSubscribed || info.metrics().contains(
+                    ClientMetricsConfigs.ALL_SUBSCRIBED_METRICS_CONFIG);
+                subscribedMetrics.addAll(info.metrics());
+                pushIntervalMs = Math.min(pushIntervalMs, info.intervalMs());
+            }
+        }
+
+        /*
+         If client matches with any subscription that has empty metrics string, then it means that client
+         is subscribed to all the metrics, so just send the empty string as the subscribed metrics.
+        */
+        if (allMetricsSubscribed) {
+            subscribedMetrics.clear();
+            // Add an empty string to indicate that all metrics are subscribed.
+            subscribedMetrics.add(ClientMetricsConfigs.ALL_SUBSCRIBED_METRICS);
+        }
+
+        int subscriptionId = computeSubscriptionId(subscribedMetrics, pushIntervalMs, clientInstanceId);
+
+        return new ClientMetricsInstance(clientInstanceId, instanceMetadata, subscriptionId, timestamp,
+            subscribedMetrics, pushIntervalMs);
+    }
+
+    /**
+     * Computes the SubscriptionId as a unique identifier for a client instance's subscription set,
+     * the id is generated by calculating a CRC32 of the configured metrics subscriptions including
+     * the PushIntervalMs, XORed with the ClientInstanceId.
+     */
+    private int computeSubscriptionId(Set<String> metrics, int pushIntervalMs, Uuid clientInstanceId) {
+        CRC32 crc = new CRC32();
+        byte[] metricsBytes = (metrics.toString() + pushIntervalMs).getBytes(StandardCharsets.UTF_8);
+        crc.update(ByteBuffer.wrap(metricsBytes));
+        return (int) crc.getValue() ^ clientInstanceId.hashCode();
+    }
+
+    private GetTelemetrySubscriptionsResponse createGetSubscriptionResponse(Uuid clientInstanceId,
+        ClientMetricsInstance clientInstance, int telemetryMaxBytes, int throttleMs) {
+
+        GetTelemetrySubscriptionsResponseData data = new GetTelemetrySubscriptionsResponseData()
+            .setClientInstanceId(clientInstanceId)
+            .setSubscriptionId(clientInstance.subscriptionId())
+            .setRequestedMetrics(new ArrayList<>(clientInstance.metrics()))
+            .setAcceptedCompressionTypes(SUPPORTED_COMPRESSION_TYPES)
+            .setPushIntervalMs(clientInstance.pushIntervalMs())
+            .setTelemetryMaxBytes(telemetryMaxBytes)
+            .setDeltaTemporality(true)
+            .setErrorCode(Errors.NONE.code())
+            .setThrottleTimeMs(throttleMs);
+
+        return new GetTelemetrySubscriptionsResponse(data);
+    }
+
+    private void validateGetRequest(GetTelemetrySubscriptionsRequest request,
+        ClientMetricsInstance clientInstance, long timestamp) {
+
+        if (!clientInstance.maybeUpdateGetRequestEpoch(timestamp) && (clientInstance.lastKnownError() != Errors.UNKNOWN_SUBSCRIPTION_ID
```

Comment:
Why only testing these two errors? What about other errors like `InvalidRequest`?

Reply:
UNKNOWN_SUBSCRIPTION_ID and UNSUPPORTED_COMPRESSION_TYPE are the only 2 exception types defined in KIP-714 which shall allow immediate retry of GetTelemetrySubscriptions request after an error from PushTelemetryRequest. For all other exception types either the API calls should be disallowed or should follow the regular push interval cycle.